solutions/usaco-971.mdx

```diff
@@ -0,0 +1,246 @@
+---
+id: usaco-971
+source: USACO Gold 2019 December
+title: Moortal Cowmbat
+author: Sofia Yang
+---
+
+[Official Analysis (C++)](http://www.usaco.org/current/data/sol_cowmbat_gold_dec19.html)
+
+## Implementation
+
+**Time Complexity:** $\mathcal{O}(M^2 + N M^2)$
+
+<LanguageSection>
+
+<CPPSection>
+
+```cpp
+CodeSnip{CPP Short Template}
+
+const int MAXN = 1e5 + 5;
+const int MAXL = 26;
+int freq[MAXN][MAXL];
+
+int count(int start, int end, int let) {
+	return freq[end + 1][let] - freq[start][let];
+}
+
+int main() {
+	setIO("cowmbat");
+	int N; int M; int K;
+	cin >> N >> M >> K;
+
+	string S;
+	cin >> S;
+
+	int cost[MAXL][MAXL];
+	for (int i = 0; i < M; i++) {
+		for (int j = 0; j < M; j++) {
+			cin >> cost[i][j];
+		}
+	}
+
+	// Calculate the minimum cost to change letter i to j.
+	for (int k = 0; k < M; k++) {
+		for (int i = 0; i < M; i++) {
+			for (int j = 0; j < M; j++) {
+				cost[i][j] = min(cost[i][j], cost[i][k] + cost[k][j]);
+			}
+		}
+	}
+
+	/* 
+	 * freq[i][j] is frequency of letter j
+	 * in the string in the range from 0 to i.
+	 */
+	for (int i = 1; i <= N; i++) {
+		for (int j = 0; j < M; j++) {
+			freq[i][j] = freq[i - 1][j];
+		}
+		freq[i][S[i - 1] - 'a']++;
+	}
+	
+	/* 
+	 * dp[i][j] is the minimum cost to make a valid streak
+	 * up to index i, where the last streak has letter j.
+	 */
+	int dp[MAXN][MAXL];
+	/*
+	 * ans[i] is the minimum cost to make a valid streak to index i.
+	 */ 
```

Comment:
no multiline

Reply:
oh wait local arrays are also kinda bad, either
1. make them global (bad)
2. use vectors instead (good)

Reply:
out of curiosity, why are local arrays bad?
and do i have to change cost, ans, etc to vectors as well?

Reply:
i would simp for you if you did

Reply:
:thonk: